apiserver/restricted_root.go

```diff
@@ -0,0 +1,38 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package apiserver
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/utils/set"
+
+	"github.com/juju/juju/rpc"
+	"github.com/juju/juju/rpc/rpcreflect"
+)
+
+// restrictedRoot restricts API calls to the environment manager and
+// user manager when accessed through the root path on the API server.
+type restrictedRoot struct {
+	rpc.MethodFinder
+}
+
+// newRestrictedRoot returns a new restrictedRoot.
+func newRestrictedRoot(finder rpc.MethodFinder) *restrictedRoot {
+	return &restrictedRoot{finder}
+}
+
+var restrictedRootNames = set.NewStrings(
+	"EnvironmentManager",
+	"UserManager",
+)
+
+// FindMethod returns a not supported error if the rootName is not one
+// of the facades available at the server root when there is no active
+// environment.
+func (r *restrictedRoot) FindMethod(rootName string, version int, methodName string) (rpcreflect.MethodCaller, error) {
+	if !restrictedRootNames.Contains(rootName) {
```

Comment:
The problem with this approach is that for completely invalid root names the client gets the same error as if they tried to access a root that's no longer available.

The other roots (e.g. upgradingRoot) do FindMethod first and let that generate a specific error for completely wrong calls, and *then* filter. Consider doing the same thing here.